docs/source/configuration.rst
Configuration
-------------

**cctl** requires some minor manual configuration to be done before you can
fully use it.

Setting Up The CCTLD User
^^^^^^^^^^^^^^^^^^^^^^^^^

**cctld** should not be run as root. Besides being insecure, it will not work.
We are best off creating a **cctld** user. It will need the **video** and
**dialout** group permissions so we can do the following:

.. code-block:: bash

   sudo useradd -r -s /bin/false cctld  # Ban it from logging in.
   sudo adduser cctld video  # Required to communicate with the video stream
   sudo adduser cctld dialout  # Required to communicate with the arduino
   sudo adduser cctld bluetooth  # Required to communicate with the bluetooth
                                 # interfaces.


.. warning::

   This assumes that there exists a ``bluetooth`` group with permissions to
   access the bluetooth interface (default on Ubuntu). If this group does not
   exist, you must create it and add the D-Bus permissions:

   .. code-block:: bash

      groupadd bluetooth
      sudo nano /etc/dbus-1/system.d/bluetooth.conf

   Then, in that file, you are to add the following policy:

   .. code-block:: xml

      <policy group="bluetooth">
        <allow send_destination="org.bluez"/>
        <allow send_interface="org.bluez.Agent1"/>
        <allow send_interface="org.bluez.GattCharacteristic1"/>
        <allow send_interface="org.bluez.GattDescriptor1"/>
        <allow send_interface="org.freedesktop.DBus.ObjectManager"/>
        <allow send_interface="org.freedesktop.DBus.Properties"/>
        <allow send_interface="org.freedesktop.DBus.Properties"/>
      </policy>

   After you do this, then you can do ``sudo adduser cctld bluetooth``.

Setting up a Proxy User
^^^^^^^^^^^^^^^^^^^^^^^

**cctld** requires a specific user it can use to setup secure proxying between **coach-os** and itself. You can create this user by running:

.. code-block:: bash

   useradd coachbot_proxy

Take note of the password you create.
It is also necessary to register the public keys of the coachbots with the
``coachbot_proxy`` user, so, you can do that with:

.. code-block:: bash

   export MY_PASS=<YOUR-PASSWORD>
   for bot in {93..103}; do
       ssh pi@192.168.1.$bot \
           "ssh-keygen -t ed25519 -N '' -f ~/.ssh/id_coachbot_proxy"
       ssh pi@192.168.1.$bot "cat ~/.ssh/id_coachbot_proxy.pub" | \
           cat > /tmp/id_coachbot_proxy.pub
       echo $MY_PASS | sudo -S -u coachbot_proxy sh -c \
           "mkdir -p /home/coachbot_proxy/.ssh; \
           cat /tmp/id_coachbot_proxy.pub >> \
           /home/coachbot_proxy/.ssh/authorized_keys"
   done
   export MY_PASS=""

We need to ensure the Coachbots can ``ssh`` proxy into the **cctld** machine,
so you should do something like:

.. code-block:: bash

   export PROXY_USER=coachbot_proxy
   cctl exec --bots=90-99 "ssh -o \"StrictHostKeyChecking=no\" \
       $PROXY_USER@192.168.1.2 'exit'"

When you're done with that, let us setup the proxy settings for **apt-get** on
the coachbots:

.. code-block:: bash

   cctl exec --bots=90-99 "echo pi | sudo -S sh -c \
      'echo Acquire::http::proxy\ \\\"socks5h://localhost:16899\\\"\; > \
          /etc/apt/apt.conf.d/12proxy'"

Copying SSH Keys
^^^^^^^^^^^^^^^^

If you were to SSH into the Coachbots, you would need to provide either an ssh
key or a password. `Providing passwords is inherently difficult to securely
achieve` and **cctl does not support password-based ssh authentication**.
Consequently, you need to copy your ssh public id to all the coachbots.
**Assuming you have a coachbot key you can use**, this task is as trivial as
running:

.. code-block:: bash

   # You can use the -n flag with ssh-copy-id for a dry run.
   for i in {3..102}; do \
       echo "Installing key to 192.168.1.$i."; \
       sshpass -p YOUR_PASSWORD -v \
           ssh-copy-id -i path/to/id_coachbot.pub "pi@192.168.1.$i"; \
   done

.. warning:: Never run this command without the ``-i`` flag. If you run the
   command without the ``-i`` flag, you will copy over all your ssh keys which
   may potentially be leaking access to the coachbots depending on how
   protected those ssh keys are.

.. note:: You can run ``ssh-copy-id`` with the ``-n`` flag to perform a dry run
   and check if everything behaves as expected before actually adding the key.

.. note:: You may encounter that this snippet will ask you about the known
   hosts for a 100 times or sshpass reporting issues with strict host checking.
   You can skip these checks with ``-o "StrictHostKeyChecking=no"`` given to
   ``ssh-copy-id``. You should, however, ensure that the IP addresses you are
   connecting to are correct. 

When you're done with that, you can simply run:

.. code-block:: bash

   for i in {3..102}; do \
       echo -n "192.168.1.$i: "; \
       ssh pi@192.168.1.$i 'echo "Successfully Connected."'; \
   done

to check if everything went as expected. You should get a printout of all
coachbots reporting a successful connection.

Udev Rules
^^^^^^^^^^

We must create some new udev rules in order for the daughterboards to create
correct Linux devices. To do this, you can simply edit
``/etc/udev/rules.d/72-cctl.rules`` (create if it doesn't exist) to
look like:

.. code-block:: text

   # Create a symlink from the rail arduino to /dev/tty-cctl-arduino
   SUBSYSTEM=="tty" ATTRS{manufacturer}=="Arduino*" \
   ATTRS{serial}=="<YOUR_SERIAL>" SYMLINK+="tty-cctl-arduino"

   # Create a symlink from the Piwebcam to /dev/video-cctl-overhead-raw
   SUBSYSTEM=="video4linux" ENV{ID_MODEL_ENC}=="Piwebcam" \
   ENV{ID_MODEL}=="Piwebcam" ENV{ID_MODEL_ID}=="0104" \
   ENV{ID_SERIAL_SHORT}=="00000000be3e8505" ENV{ID_TYPE}=="video" \
   SYMLINK+="video71" \
   RUN{program}+="/sbin/modprobe v4l2loopback video_nr=72"

   # Create a processed stream whenever the arduino is connected.

To find the serial number, you can run ``udevadm info -a -n /dev/ttyACM<X>``
where ``<X>`` is the number that you are certain the Arduino is on.

You can plug out and then plug back in the Arduino daughterboard.

Now that you've done this, you are guaranteed to have ``/dev/tty-cctl-arduino``
as a file (as a symlink to  ``/dev/ttyACM*``).

Configuration Files
^^^^^^^^^^^^^^^^^^^

**cctl** exposes some configuration files that you can use to tweak its
behavior. These configuration files are normally located either in
``~/.config/coachswarm/`` or ``/etc/coachswarm/``.

.. note:: **cctl** reads ``~/.config/coachswarm/`` configuration files first
   and falls back to ``/etc/coachswarm/`` files if those do not exist. If
   **cctl** is unable to find either of these files, it will automatically
   generate new files in ``~/.config/coachswarm/{cctl,coachswarm}.conf`` with
   sane defaults.

From now, the term ``$CONFDIR`` refers to the loaded configuration directory as
described. In ``$CONFDIR`` two configuration files are supported.

cctl.conf
^^^^^^^^^

The ``$CONFDIR/cctl.conf`` file configures **cctl**’s behavior. The following
is an example file with all supported keys.

.. include:: ../../src/cctl_static/cctl.conf
   :code: ini
   :name: cctld.conf

Configuring **cctld**
^^^^^^^^^^^^^^^^^^^^^

You are required to configure **cctld** in order to achieve commuincation
between the coachbots and it. The configuration file should be stored in
``/etc/coachswarm/cctld.conf`` and **cctld** will attempt to read from it. If
this file does not exist, **cctld** will **not start**. The format of the file
is:

.. include:: ../../src/cctl_static/cctld.conf
   :code: ini
   :name: cctld.conf
   
Systemd Service
~~~~~~~~~~~~~~~

To get **cctld** to be a proper Linux service we have many options. We could
put a script in ``/etc/init.d/`` (and if you are more familiar with that you
are more than welcome to). However, I personally prefer to use
**systemd** [#fsystemd]_. To do this, we need to make a simple file to let
**systemd** know about **cctld** and let it manage it as a proper service.
Something along the lines of the following should work.

.. include:: ../../src/cctl_static/cctld.service
    :code:
    :name: /etc/systemd/system/cctld.service

Then you can tell **systemd** to enable this service on boot and immediately
start it:

.. code-block:: bash

   sudo systemctl enable cctld
   sudo systemctl start cctld

The next time your computer boots up, **cctld** will start automatically. You
don't even need to log in.

When you're done configuring **cctl**, you can visit `Usage <usage.html>`_ for
information on how to effectively use **cctl**.

.. rubric:: Footnotes

.. [#fsystemd] I know that **systemd** has its flaws and that it is not the
   prefered tool of many users. **cctld** is written as daemon and not as a
   **systemd** service for this very reason. You are by no means obliged to use
   **systemd** if you are a fan of traditional service management.


src/cctl/api/configuration.py
# -*- coding: utf-8 -*-

"""
Singleton module containing configuration values.
"""

import sys
from os import makedirs, path
import logging
from configparser import ConfigParser

from cctl.res import ERROR_CODES, RES_STR

usr_conf_dir = path.join(path.expanduser('~'), '.config', 'coachswarm')

conf_file_paths = [
    path.join(usr_conf_dir, 'cctl.conf'),
    path.join('/', 'etc', 'coachswarm', 'cctl.conf'),
]

mandatory_keys = {
    'server': [
        'interface',
        'path'
    ],
    'coachswarm': [
        'conf_path',
        'remote_path',
        'id_range_min',
        'id_range_max',
        'ssh_user',
        'ssh_key',
        'net_server_port_pub',
        'net_server_port_rep',
        'net_server_port_req',
        'socks5_port',
        'socks5_proxy_user',
        'pi_password'
    ],
    'logs': [
        'syslog_path',
        'legacy_log_file_path'
    ],
    'cctld': [
        'state_feed',
        'request_feed'
    ]
}

# TODO: These should be replaced with a copy from cctl_static/cctl.conf
default_values = {
    'server': {
        'interface': 'enp60s0',
        'path': '/home/hanlin/coach/server_beta'
    },
    'coachswarm': {
        'conf_path': path.join(usr_conf_dir, 'coachswarm.conf'),
        'remote_path': '/home/hanlin/control',
        'id_range_min': 0,
        'id_range_max': 99,
        'ssh_user': 'pi',
        'ssh_key': path.join(path.expanduser('~'), '.ssh', 'id_coachbot'),
        'net_server_port_rep': 16891,
        'net_server_port_pub': 16892,
        'net_server_port_req': 16893,
        'socks5_port': 16899,
        'socks5_proxy_user': 'coachbot_proxy',
        'pi_password': 'pi'  # TODO: This is so insecure. Problem is, there is
                             # no dedicated linux user running on the
                             # coachbots.
    },
    'logs': {
        'syslog_path': '/var/log/syslog',
        'legacy_log_file_path': '/home/pi/control/experiment_log'
    },
    'cctld': {
        'state_feed': 'ipc:///run/cctld/state_feed',
        'request_feed': 'ipc:///run/cctld/feed_pipe'
    }
}

config = ConfigParser()

# Attempt to read config if possible.
files_found = config.read(conf_file_paths)

if len(files_found) == 0:
    logging.warning(RES_STR['conf_error'])
    config = ConfigParser()

    config.update(default_values)

    if not path.exists(usr_conf_dir):
        makedirs(usr_conf_dir)

    # Write cctl.conf
    with open(path.join(usr_conf_dir, 'cctl.conf'), 'w') as conf_file:
        config.write(conf_file)

    # Write coachswarm.conf
    cs_conf_path = path.join(usr_conf_dir, 'coachswarm.conf')
    if not path.exists(cs_conf_path):
        with open(cs_conf_path, 'w') as conf_file:
            conf_file.write('{\n"COM_RANGE": 100.0\n}')


def _exit_on_err_key(m_key: str):
    logging.error(RES_STR['conf_malformed'], m_key)
    sys.exit(ERROR_CODES['conf_error'])


for key, values in mandatory_keys.items():
    if key not in config:
        _exit_on_err_key(key)

    for value in values:
        if value not in config[key]:
            _exit_on_err_key(value)


def get_server_dir():
    """
    Returns:
        The operating server directory.
    """
    return config['server']['path']


def get_server_interface():
    """
    Returns:
        The server interface
    """
    return config['server']['interface']


def get_coachswarm_conf_path():
    """
    Returns:
        The coachswarm.conf filepath
    """
    return config['coachswarm']['conf_path']


def get_syslog_path() -> str:
    """Returns the syslog path.

    Returns:
        str: The user-configured syslog path.
    """
    return config.get('logs', 'syslog_path')


def get_legacy_log_file_path() -> str:
    """Returns the legacy log file name.

    This filename is the filename cctl expects legacy logs to be called on the
    coachbots.

    Returns:
        str: The user configured legacy log filename.
    """
    return config.get('logs', 'legacy_log_file_path')


def get_coachswarm_ssh_user() -> str:
    """Returns the ssh user to use for the coachbots."""
    return config.get('coachswarm', 'ssh_user')


def get_coachswarm_path_to_ssh_key() -> str:
    """Returns the path to the coachbot ssh key."""
    return path.expanduser(config.get('coachswarm', 'ssh_key'))


def get_valid_coachbot_range():
    """Returns the valid Coachbot range."""
    return range(
        config.getint('coachswarm', 'id_range_min'),
        config.getint('coachswarm', 'id_range_max') + 1
    )


def get_coachswarm_net_rep_port() -> int:
    """Returns the port used for the networking with the coachbots on the REP
    transport."""
    return config.getint('coachswarm', 'net_server_port_rep')


def get_coachswarm_net_pub_port() -> int:
    """Returns the port used for networking with the coachbots on the PUB
    transport."""
    return config.getint('coachswarm', 'net_server_port_pub')


def get_coachswarm_net_req_port() -> int:
    """Returns the port used for networking with the coachbots on the REQ
    transport."""
    return config.getint('coachswarm', 'net_server_port_req')


def get_coachswarm_remote_path() -> str:
    """Returns the path to the remote directory."""
    return config.get('coachswarm', 'remote_path')


def get_socks5_port() -> int:
    """Returns the default socks5 port used when creating a proxy to cctl."""
    return config.getint('coachswarm', 'socks5_port')


def get_pi_password() -> str:
    """Returns the pi user password. This is insecure, but a necessary evil
    because code is executed as root on the legacy implementation.
    """
    return config.get('coachswarm', 'pi_password')


def get_socks5_proxy_user() -> str:
    """Returns the cctl user setup for proxying the internet."""
    return config.get('coachswarm', 'socks5_proxy_user')


def get_state_feed() -> str:
    """Returns the full path to the cctld state feed."""
    return config.get('cctld', 'state_feed')


def get_request_feed() -> str:
    """Returns the request feed of cctld."""
    return config.get('cctld', 'request_feed')
